pkg/kube/dialer.go

```diff
@@ -0,0 +1,336 @@
+// matches Kubernetes' own DNS records and a pod within the cluster would be able
+// to resolve the same FQDN to the pod that we do.
+//
+// These assumptions allow us to use this dialer to issues requests *as if* we are
+// in the Kubernetes network even from outside of it (i.e. in tests that attempt to
+// connect to a pod at a given DNS address).
+func (p *PodDialer) parseDNS(fqdn string) (pod types.NamespacedName, port int, err error) {
+	addressPort := strings.Split(fqdn, ":")
+	if len(addressPort) != 2 {
+		err = ErrNoPort
+		return
+	}
+
+	port, err = strconv.Atoi(addressPort[1])
+	if err != nil {
+		return
+	}
+
+	fqdn = addressPort[0]
+	fqdn = strings.TrimSuffix(fqdn, "."+p.clusterDomain)
+	fqdn = strings.TrimSuffix(fqdn, "."+svcLabel)
+	labels := strings.Split(fqdn, ".")
+
+	// since we only dial pods we require 2 labels
+	// (assuming we're trying to dial a pod in the
+	// default namespace) or 3 (for a pod outside
+	// of the default namespace)
+	switch len(labels) {
+	case 2:
+		pod.Namespace = defaultNamespace
+	case 3:
+		pod.Namespace = labels[2]
+	default:
+		err = ErrInvalidPodFQDN
+		return
+	}
+
+	// service - labels[1]
+	pod.Name = labels[0]
+	return
+}
+
+// mutex must be held while calling this
+func (p *PodDialer) connectionForPod(pod types.NamespacedName) (httpstream.Connection, error) {
+	if conn, ok := p.connections[pod]; ok {
+		return conn, nil
+	}
+
+	transport, upgrader, err := spdy.RoundTripperFor(p.config)
+	if err != nil {
+		return nil, err
+	}
+
+	cfg := p.config
+	cfg.APIPath = "/api"
+	cfg.GroupVersion = &schema.GroupVersion{Version: "v1"}
+	cfg.NegotiatedSerializer = negotiatedSerializer
+	restClient, err := rest.RESTClientFor(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	req := restClient.Post().
+		Resource("pods").
+		Namespace(pod.Namespace).
+		Name(pod.Name).
+		SubResource("portforward")
+
+	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: transport}, http.MethodPost, req.URL())
+	conn, protocol, err := dialer.Dial(portForwardProtocolV1Name)
+	if err != nil {
+		return nil, err
+	}
+	if protocol != portForwardProtocolV1Name {
+		return nil, fmt.Errorf("unable to negotiate protocol: client supports %q, server returned %q", portForwardProtocolV1Name, protocol)
+	}
+
+	p.connections[pod] = conn
+	return conn, nil
+}
+
+type conn struct {
+	httpstream.Stream
+	errorStream httpstream.Stream
+	network     string
+	remote      string
+
+	errCh chan error
+}
+
+func wrapConn(network, remote string, s, err httpstream.Stream) *conn {
+	c := &conn{
+		Stream:      s,
+		errorStream: err,
+		network:     network,
+		remote:      remote,
+		errCh:       make(chan error, 1),
+	}
+
+	go c.pollErrors()
+	return c
+}
+
+func (c *conn) pollErrors() {
+	data, err := io.ReadAll(c.errorStream)
+	if err != nil {
+		c.writeError(err)
+		return
+	}
+
+	if len(data) != 0 {
+		c.writeError(errors.New(string(data)))
+		return
+	}
+}
+
+func (c *conn) writeError(err error) {
+	select {
+	case c.errCh <- err:
```

Comment:
Why is there a `default` here? `errCh` is buffered with 1 element and it appears to only ever be written to once. The default could lead to us unexpectedly dropping errors. Given that the `DialContext` doesn't support cancelation, we _really_ shouldn't be dropping errors even if this is test only code.

Reply:
So, the idea is that an error in the underlying `errorStream` poller would cause the connection to terminate, so we really only care about the first error ever returned. That said, I'm cleaning up some of this to more directly indicate that the connection is no longer valid once we've received some sort of error and that everything should be trashed.

The reason for the buffered channel is to intentionally drop anything but the first write. You're right about the error poller only having written to the channel once initially, missed that 😬. I added a context checking goroutine which will also write here though, so I think I'll likely just keep this as is for now until you have a second pass review.